src/utils/elections/states/pennsylvania.py

```diff
@@ -0,0 +1,188 @@
+    "Content-Type": "application/json",
+    "Origin": "https://www.electionreturns.pa.gov",
+    "Referer": "https://www.electionreturns.pa.gov/ReportCenter/Reports",
+}
+TOO_MANY_REQUESTS = 429
+
+
+def fetch_election_data(election_id: int, subtype: str) -> dict:
+    """Fetch election data for a given election ID and subtype.
+
+    Args:
+        election_id: The ID of the election to fetch data for.
+        subtype: The subtype of the election to fetch data for.
+            "G" for General, "P" for Primary, "S" for Special.
+
+    Returns:
+        A pandas DataFrame containing the election data.
+    """
+    payload = {
+        "ElectionID": election_id,
+        "ElectionsubType": subtype,
+        "ExportType": "C",
+        "ReportType": "S",
+        "FileName": "Official",
+    }
+    response = requests.post(BASE_URL, headers=HEADERS, json=payload, timeout=10)
+    if response.status_code == TOO_MANY_REQUESTS:
+        print("Rate limited. Retrying after a delay...")
+        time.sleep(10)
+        return fetch_election_data(election_id, subtype)
+    response.raise_for_status()
+    # Convert the response text to a pandas DataFrame
+    csv_data = StringIO(response.text)
+    election_details = pd.read_csv(csv_data)
+    return election_details
+
+
+def fetch_all_results(elections: list[dict]) -> pd.DataFrame:
+    """Fetch results for all valid elections.
+
+    Args:
+        elections: A list of dictionaries containing election IDs and subtypes.
+
+    Returns:
+        A dataframe containing the election results for all valid elections.
+    """
+    results = []
+    for election in elections:
+        print(
+            f"Fetching results for ElectionID {election['ElectionID']} Subtype {election['Subtype']}..."
+        )
+        try:
+            election_df = fetch_election_data(
+                election["ElectionID"], election["Subtype"]
+            )
+            election_df.loc[:, "election--year"] = election["Year"]
+            results.append(election_df)
+        except Exception as e:
+            print(f"Failed to fetch data for ElectionID {election['ElectionID']}: {e}")
+        time.sleep(1)  # Avoid hitting rate limits
+    return pd.concat(results, ignore_index=True)
+
+
+def get_all_election_ids(start_year: int = None, end_year: int = None) -> list[dict]:
+    """Get all valid election IDs and subtypes using the API.
+
+    Args:
+        start_year: The starting year for the elections to fetch.
+            If None, defaults to the earliest year available.
+        end_year: The ending year for the elections to fetch.
+            If None, defaults to the latest year available.
+
+    Returns:
+        A list of dictionaries containing: "EclectionID", "Subtype", and "Year".
+        "ElectionID" and "Subtype" are used to fetch election data.
+    """
+    url = "https://www.electionreturns.pa.gov/api/ElectionReturn/GetAllElections?methodName=GetAllElections"
```

Comment:
probably better to have this URL and headers for the election IDs as static variables at the top as well, for consistency

Reply:
Good catch, not sure how I missed that